tests/e2e/main.go

```diff
@@ -287,226 +283,121 @@ func getTestCases(selectedPredefinedTests, selectedTestFiles TestSet) (tests []t
-		useGaia = "true"
-		if tr.gaiaTag != "" {
-			majVersion, err := strconv.Atoi(tr.gaiaTag[1:strings.Index(tr.gaiaTag, ".")])
-			if err != nil {
-				panic(fmt.Sprintf("invalid gaia version %s", tr.gaiaTag))
-			}
-			if majVersion < 9 {
-				panic(fmt.Sprintf("gaia version %s is not supported - supporting only v9.x.x and newer", tr.gaiaTag))
+func createTestRunners(targets []ExecutionTarget, testCases []testStepsWithConfig) []TestRunner {
+	runners := []TestRunner{}
+	for _, target := range targets {
+		for _, tc := range testCases {
+			tr := TestRunner{
+				config:  tc.testRun,
+				steps:   tc.steps,
+				target:  target,
+				verbose: *verbose,
 			}
-			gaiaTag = tr.gaiaTag
+			//TODO: refactor this target specific setting
+			tr.target.(*DockerContainer).containerCfg = tc.testRun.containerConfig
+			tr.config.transformGenesis = *transformGenesis
+			tr.config.SetCometMockConfig(*useCometmock)
+			tr.config.SetRelayerConfig(*useGorelayer)
+			runners = append(runners, tr)
 		}
 	}
-	scriptStr := fmt.Sprintf(
-		"tests/e2e/testnet-scripts/start-docker.sh %s %s %s %s %s",
-		tr.containerConfig.ContainerName,
-		tr.containerConfig.InstanceName,
-		localSdk,
-		useGaia,
-		gaiaTag,
-	)
-
-	//#nosec G204 -- Bypass linter warning for spawning subprocess with cmd arguments.
-	cmd := exec.Command("/bin/bash", "-c", scriptStr)
-
-	cmdReader, err := cmd.StdoutPipe()
-	if err != nil {
-		log.Fatal(err)
-	}
-	cmd.Stderr = cmd.Stdout
+	return runners
+}
 
-	if err := cmd.Start(); err != nil {
-		log.Fatal(err)
+func executeTests(runners []TestRunner) error {
+	if parallel != nil && *parallel {
+		fmt.Println("=============== running all tests in parallel ===============")
 	}
 
-	scanner := bufio.NewScanner(cmdReader)
+	var wg sync.WaitGroup
+	var err error = nil
 
-	for scanner.Scan() {
-		out := scanner.Text()
-		if verbose != nil && *verbose {
-			fmt.Println("startDocker: " + out)
-		}
-		if out == "beacon!!!!!!!!!!" {
-			return
+	for _, runner := range runners {
+		if parallel != nil && *parallel {
+			wg.Add(1)
+			go func(runner TestRunner) {
+				defer wg.Done()
+				runner.Run()
+			}(runner)
+		} else {
+			fmt.Printf("=============== running %s ===============\n", runner.config.name)
+			err = runner.Run()
 		}
 	}
-	if err := scanner.Err(); err != nil {
-		log.Fatal(err)
+
+	if parallel != nil && *parallel {
+		wg.Wait()
 	}
 
-	err = cmd.Wait()
-	log.Fatalf("StartDocker exited with error: %v", err)
+	return err
 }
 
-// remove docker container to reduce resource usage
-// otherwise the chain will keep running in the background
-func (tr *TestConfig) teardownDocker() {
-	fmt.Printf("===============  tearing down %s testRun ===============\n", tr.name)
-	//#nosec G204 -- Bypass linter warning for spawning subprocess with cmd arguments.
-	cmd := exec.Command("docker", "kill", tr.containerConfig.InstanceName)
+// runs E2E tests
+// all docker containers are built sequentially to avoid race conditions when using local cosmos-sdk
```

Comment:
Not sure I get this - would parallel builds really have a race condition? Could you explain a bit more why this is the case?